### PR TITLE
root: make calling `Close` on the root actually close everything

### PR DESCRIPTION
**Pull request by Stebalien**
That way, we can close and *then* cancel the context as a final teardown. Really, we should consider not passing a context in the first place but it *is* useful for requests (where we just want to abort).

### CODE REVIEW

File: root.go [hunk 1]
```diff
@@ -85,19 +85,23 @@ func IsFile(fsn FSNode) bool {
...
 // Root represents the root of a filesystem tree.
 type Root struct {
-
 	// Root directory of the MFS layout.
 	dir *Directory
 
 	repub *Republisher
+
+	ctx    context.Context
+	cancel context.CancelFunc
```
Comment (gammazero) [inline root.go:94]: I think this is a misuse of a context.
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.

-- [golang docs](https://golang.org/pkg/context/)

unless the struct is used as a parameter, according to Brad Fitzpatrick

It also makes it confusing about what is the correct way to shutdown msf.  Call `Close`, cancel the context, both?  Better to have one or the other.  How about using a context _only_ to cancel a function call that may take too long to return, and using `Close` to shutdown mfs root.

File: root.go [hunk 2]
```diff
@@ -190,7 +196,7 @@ func (kr *Root) FlushMemFree(ctx context.Context) error {
 // the top), document it and maybe make it an anonymous variable (if
 // that's possible).
 func (kr *Root) updateChildEntry(c child) error {
-	err := kr.GetDirectory().dagService.Add(context.TODO(), c.Node)
+	err := kr.GetDirectory().dagService.Add(kr.ctx, c.Node)
```
Comment (gammazero) [inline root.go:199]: It seems like the context here should be passed into this function as an argument, and used to cancel _this call_ should it take too long or need to be canceled.  I guess this is to cause this `dagService.Add()` to fail if this mfs root has been closed, but adds to my confusion about calling `Close` vs canceling the context.

**Review by gammazero — COMMENTED** [review]
The PR is an improvement, but it continues on with what I think is a misuse of context.  See comments.

However, if my concern is misplaced or requires changes outside the scope of this PR, then this should probably get merged.

**Comment by aschmahmann** [comment]
This repository is no longer maintained and has been copied over to [Boxo](https://github.com/ipfs/boxo).  In an effort to avoid noise and crippling in the Boxo repo from the weight of issues of the past, we are closing most issues and PRs in this repo.  Please feel free to open a new issue in Boxo (and reference this issue) if resolving this issue is still critical for unblocking or improving your usecase.

You can learn more in the [FAQs for the Boxo repo copying/consolidation effort](https://github.com/ipfs/boxo/wiki/Copied-or-Migrated-Repos-FAQ).

**Pull request closed without merging**